modules/simulation/DCSupplySimulator/main/power_supply_DCImpl.cpp

```diff
@@ -0,0 +1,117 @@
+    this->connector_voltage = 0.0;
+    this->connector_current = 0.0;
+
+    this->power_supply_handler = std::thread(&power_supply_DCImpl::power_supply_worker, this);
+}
+
+void power_supply_DCImpl::ready() {
+}
+
+types::power_supply_DC::Capabilities power_supply_DCImpl::handle_getCapabilities() {
+    types::power_supply_DC::Capabilities Capabilities = {
+        .bidirectional = config.bidirectional,
+        .current_regulation_tolerance_A = 2.0,
+        .peak_current_ripple_A = 2.0,
+        .max_export_voltage_V = static_cast<float>(config.max_voltage),
+        .min_export_voltage_V = static_cast<float>(config.min_voltage),
+        .max_export_current_A = static_cast<float>(config.max_current),
+        .min_export_current_A = static_cast<float>(config.min_current),
+        .max_export_power_W = static_cast<float>(config.max_power),
+        .max_import_voltage_V = static_cast<float>(config.max_voltage),
+        .min_import_voltage_V = static_cast<float>(config.min_voltage),
+        .max_import_current_A = static_cast<float>(config.max_current),
+        .min_import_current_A = static_cast<float>(config.min_current),
+        .max_import_power_W = static_cast<float>(config.max_power),
+        .conversion_efficiency_import = 0.85,
+        .conversion_efficiency_export = 0.9,
+    };
+    return Capabilities;
+}
+
+void power_supply_DCImpl::handle_setMode(types::power_supply_DC::Mode& value) {
+    this->mode = value;
+
+    if ((value == types::power_supply_DC::Mode::Off) || (value == types::power_supply_DC::Mode::Fault)) {
+        this->connector_voltage = 0.0;
+        this->connector_current = 0.0;
+    } else if (value == types::power_supply_DC::Mode::Export) {
+        this->connector_voltage = this->settings_connector_export_voltage;
+        this->connector_current = this->settings_connector_max_export_current;
+    } else if (value == types::power_supply_DC::Mode::Import) {
+        this->connector_voltage = this->settings_connector_import_voltage;
+        this->connector_current = this->settings_connector_max_import_current;
+    }
+
+    mod->p_main->publish_mode(value);
+}
+
+void power_supply_DCImpl::clampVoltageCurrent(double& voltage, double& current) {
+    voltage = voltage < this->config.min_voltage   ? this->config.min_voltage
+              : voltage > this->config.max_voltage ? this->config.max_voltage
+                                                   : voltage;
+
+    current = current < this->config.min_current   ? this->config.min_current
+              : current > this->config.max_current ? this->config.max_current
+                                                   : current;
+}
+
+void power_supply_DCImpl::handle_setExportVoltageCurrent(double& voltage, double& current) {
+    double temp_voltage = voltage;
+    double temp_current = current;
+
+    clampVoltageCurrent(temp_voltage, temp_current);
+
+    std::lock_guard<std::mutex> access_lock(this->power_supply_values_mutex);
```

Comment:
With C++17 you could also use "std::scoped_lock access_lock(this->power_supply_values_mutex)" with imho is a little bit more readable as its name indicates that is is valid for the scope. But just a matter of taste.